### PR TITLE
Homework node.js w2

### PR DESCRIPTION
**Pull request by myshuker**

### CODE REVIEW

File: nodejs/week2/Meal_sharing_website/src/backend/index.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
+const express = require("express");
+const app = express();
+
+const getAllMeals = require("./routes/meals");
+const getAllReservations = require("./routes/reservations");
+const getAllReviews = require("./routes/reviews");
+ 
+// create middleware
+app.use((req, res, next) => {
```
Comment (Hatko) [inline nodejs/week2/Meal_sharing_website/src/backend/index.js]: You can try to use the second answer here:

https://stackoverflow.com/questions/3552461/how-to-format-a-javascript-date

File: nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+function getMeals(req, res) {
+  const idString = req.params.id;
+  const id = parseInt(idString);
+  const max_price = parseInt(req.query.maxPrice);
+  const title = req.query.title;
+  const createdAfter = req.query.createdAfter;
+  const limit = parseInt(req.query.limit);
+
+  if (id) {
+    if (id === 0 || id > jsonMeals.length)
```
Comment (Hatko) [inline nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js]: You will still call `res.json(data)` if this condition is true. Probably you want to return after the error occurred. 

This is not important and doesn't have to be changed, but it's a little bit weak logic in general to assume, that you:
1. Will never have an item with id `0`.
2. Will never have an item larger than the number of elements in the array.

File: nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+      res.status(404).send(`The item with id: ${id} was not found`);
+    const data = jsonMeals.filter(item => {
+      if (item.id === id) {
```
Comment (Hatko) [inline nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js]: 1. `filter` function has to return boolean (it returns an object here)
2. `filter` function has to always return a value (it doesn't return anything - returns undefined implicitly - if `item.id !== id`

Try to fix both problems, and if you don't manage to - ask me next time in the class 😉

File: nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        return item;
+      }
+    });
+
+    // console.log(data);
+    res.json(data);
+  } else if (idString) {
+    // console.log(typeof idString);
+    res.send(`Id it should be a number`);
```
Comment (Hatko) [inline nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js]: You could also use status here 😉 

Just a suggestion though.

File: nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    res.json(meal_title);
+  } else if (createdAfter) {
+    const meal_created_after = jsonMeals.filter(item => {
+      return item.createdAt > createdAfter;
+    });
+    res.json(meal_created_after);
+  } else if (limit) {
+    const meal_limit = jsonMeals.slice(limit);
+    res.json(meal_title);
+  } else if (!id) {
```
Comment (Hatko) [inline nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js]: The check will always be `true`, as you already checked the opposite as the first `if`.

File: nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+      return item.title.toLowerCase() === title.toLowerCase();
+    });
+    res.json(meal_title);
+  } else if (createdAfter) {
+    const meal_created_after = jsonMeals.filter(item => {
+      return item.createdAt > createdAfter;
+    });
+    res.json(meal_created_after);
+  } else if (limit) {
+    const meal_limit = jsonMeals.slice(limit);
```
Comment (Hatko) [inline nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js]: The app crashes if you send `limit` as a query parameter. 

After you fix it, I think it won't still work as expected. Try to test it.

File: nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    res.json(meal_title);
+  } else if (!id) {
+    res.json(jsonMeals);
+  }
+}
+
+router.get("/", getMeals);
```
Comment (Hatko) [inline nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js]: I think you don't really need this line. The next one will take care of both `/` and `/:id`.

File: nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+router.get("/:id", getMeals);
+
+module.exports = router;
+
+//function getMeals(req, res) {
```
Comment (Hatko) [inline nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js]: Has to probably be removed 🙂

File: nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js (hunk 3 above)
Comment (Hatko) [reply nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js]: The same pattern repeats in several places.

File: nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js (hunk 7 above)
Comment (Hatko) [reply nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js]: The same pattern repeats in several places.

File: nodejs/week2/Meal_sharing_website/src/backend/index.js (hunk 1 above)
Comment (myshuker) [reply nodejs/week2/Meal_sharing_website/src/backend/index.js]: hello Hatko
do you mean this method :+1: const  today  = new Date();
  console.log(today.toLocaleString("en-US"));

File: nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js (hunk 2 above)
Comment (myshuker) [reply nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js]: so you mean I should write it in this way : 
if (id) {
    // if (id === 0 || id > jsonMeals.length)
    //   res.status(404).send(`The item with id: ${id} was not found`);
    const data = jsonMeals.filter(item => {
      if (item.id === id) {
        return item;
      }
    });

File: nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js (hunk 4 above)
Comment (myshuker) [reply nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js]: witch status number you mean ?

File: nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js (hunk 6 above)
Comment (myshuker) [reply nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js]: do you mean should I write it as : 
    const meal_limit = jsonMeals.slice(0, limit);

Comment (myshuker) [reply nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js]: also its still not works

File: nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js (hunk 7 above)
Comment (myshuker) [reply nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js]: when I commet it I get  : 
Cannot GET /meals

File: nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js (hunk 8 above)
Comment (myshuker) [reply nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js]: :+1:

File: nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js (hunk 3 above)
Comment (myshuker) [reply nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js]: :+1:

Comment (myshuker) [reply nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js]: is this ok : 
if (id) {
        const data = jsonMeals.filter(item => {
        return  item.id === id
        });
    res.json(data);
  } else if (idString) {
    res.send(`Id it should be a number`);
  } else if (max_price) {
    const dataPrice = jsonMeals.filter(item => {
      return item.price < max_price;
    });
    res.json(dataPrice);
  } else if (title) {
    const meal_title = jsonMeals.filter(item => {
      return item.title.toLowerCase() === title.toLowerCase();
    });
    res.json(meal_title);
  } else if (createdAfter) {
    const meal_created_after = jsonMeals.filter(item => {
      return item.createdAt > createdAfter;
    });
    res.json(meal_created_after);
  } else if (limit) {
    const meal_limit = jsonMeals.slice(0, limit);
    res.json(meal_title);
  } else if (!id) {
    res.json(jsonMeals);
  }
}

File: nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js (hunk 4 above)
Comment (myshuker) [reply nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js]: you mean : 
res.status(400)

Comment (myshuker) [reply nodejs/week2/Meal_sharing_website/src/backend/routes/meals.js]: res.status(400).send(`Id it should be a number`);

**Pull request merged**